File: src/ultidraft/domain/notes.py
"""Sidecar notes stored beside a manuscript. No Qt — Android v2 writes this shape."""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from pathlib import Path

from ultidraft.domain.lexicon import LexiconRule

SIDECAR_VERSION = 1


@dataclass
class Note:
    id: str
    created: str
    chapter_title: str
    anchor_quote: str
    context_before: str
    context_after: str
    sentence_index: int
    body: str

    def to_dict(self) -> dict:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict) -> Note:
        return cls(
            id=str(data["id"]),
            created=str(data["created"]),
            chapter_title=str(data.get("chapter_title", "")),
            anchor_quote=str(data.get("anchor_quote", "")),
            context_before=str(data.get("context_before", "")),
            context_after=str(data.get("context_after", "")),
            sentence_index=int(data.get("sentence_index", 0)),
            body=str(data.get("body", "")),
        )


@dataclass
class Sidecar:
    version: int
    manuscript_path: str
    manuscript_hash: str
    position: dict
    notes: list[Note] = field(default_factory=list)
    lexicon: list[LexiconRule] = field(default_factory=list)

    def next_note_id(self) -> str:
        highest = 0
        for note in self.notes:
            match = note.id[1:] if note.id.startswith("N") else note.id
            try:
                highest = max(highest, int(match))
            except ValueError:
                continue
        return f"N{highest + 1:03d}"

    def add_note(self, note: Note) -> None:
        self.notes.append(note)

    def to_dict(self) -> dict:
        return {
            "version": self.version,
            "manuscript_path": self.manuscript_path,
            "manuscript_hash": self.manuscript_hash,
            "position": self.position,
            "notes": [note.to_dict() for note in self.notes],
            "lexicon": [rule.to_dict() for rule in self.lexicon],
        }

    @classmethod
    def new(cls, manuscript_path: str, manuscript_hash: str) -> Sidecar:
        return cls(
            version=SIDECAR_VERSION,
            manuscript_path=manuscript_path,
            manuscript_hash=manuscript_hash,
            position={"chapter_id": "", "sentence_index": 0},
            notes=[],
            lexicon=[],
        )

    @classmethod
    def from_dict(cls, data: dict) -> Sidecar:
        notes = [Note.from_dict(item) for item in data.get("notes", [])]
        lexicon = [LexiconRule.from_dict(item) for item in data.get("lexicon", [])]
        position = data.get("position") or {"chapter_id": "", "sentence_index": 0}
        return cls(
            version=int(data.get("version", SIDECAR_VERSION)),
            manuscript_path=str(data.get("manuscript_path", "")),
            manuscript_hash=str(data.get("manuscript_hash", "")),
            position={
                "chapter_id": str(position.get("chapter_id", "")),
                "sentence_index": int(position.get("sentence_index", 0)),
            },
            notes=notes,
            lexicon=lexicon,
        )


def sidecar_path(manuscript_path: Path) -> Path:
    return manuscript_path.with_name(f"{manuscript_path.stem}.ultidraft.json")


def load_sidecar(manuscript_path: Path, manuscript_hash: str) -> Sidecar:
    path = sidecar_path(manuscript_path)
    if not path.is_file():
        return Sidecar.new(manuscript_path.name, manuscript_hash)
    data = json.loads(path.read_text(encoding="utf-8"))
    sidecar = Sidecar.from_dict(data)
    sidecar.manuscript_path = manuscript_path.name
    return sidecar


def save_sidecar(manuscript_path: Path, sidecar: Sidecar) -> Path:
    path = sidecar_path(manuscript_path)
    path.write_text(
        json.dumps(sidecar.to_dict(), indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )
    return path


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).astimezone().isoformat(timespec="seconds")

File: android/app/src/main/java/com/ultidraft/domain/Notes.kt
package com.ultidraft.domain

import java.time.OffsetDateTime
import java.time.temporal.ChronoUnit

/**
 * The sidecar stored beside a manuscript, byte-compatible with the desktop app.
 *
 * Phone-made notes are numbered `M001`, `M002`, … while the desktop numbers `N001`, …
 * The desktop's `next_note_id` ignores ids it cannot parse as `N<int>`, so the two
 * clients can both add notes to a Syncthing-synced book without ever colliding on an id.
 */

const val SIDECAR_VERSION = 1
const val MOBILE_NOTE_PREFIX = "M"
const val DESKTOP_NOTE_PREFIX = "N"

data class Note(
    val id: String,
    val created: String,
    val chapterTitle: String,
    val anchorQuote: String,
    val contextBefore: String,
    val contextAfter: String,
    val sentenceIndex: Int,
    val body: String,
) {
    fun toJson(): JsonValue = JsonValue.obj(
        "id" to JsonValue.str(id),
        "created" to JsonValue.str(created),
        "chapter_title" to JsonValue.str(chapterTitle),
        "anchor_quote" to JsonValue.str(anchorQuote),
        "context_before" to JsonValue.str(contextBefore),
        "context_after" to JsonValue.str(contextAfter),
        "sentence_index" to JsonValue.num(sentenceIndex),
        "body" to JsonValue.str(body),
    )

    companion object {
        fun fromJson(value: JsonValue): Note = Note(
            id = value.string("id"),
            created = value.string("created"),
            chapterTitle = value.string("chapter_title"),
            anchorQuote = value.string("anchor_quote"),
            contextBefore = value.string("context_before"),
            contextAfter = value.string("context_after"),
            sentenceIndex = value.int("sentence_index"),
            body = value.string("body"),
        )
    }
}

data class Position(val chapterId: String, val sentenceIndex: Int)

data class Sidecar(
    val version: Int = SIDECAR_VERSION,
    var manuscriptPath: String,
    var manuscriptHash: String,
    var position: Position = Position("", 0),
    val notes: MutableList<Note> = mutableListOf(),
    var lexicon: List<LexiconRule> = emptyList(),
) {
    /** The next free phone-side id, e.g. `M003`. */
    fun nextNoteId(prefix: String = MOBILE_NOTE_PREFIX): String {
        var highest = 0
        for (note in notes) {
            if (!note.id.startsWith(prefix)) continue
            val number = note.id.removePrefix(prefix).toIntOrNull() ?: continue
            if (number > highest) highest = number
        }
        return prefix + (highest + 1).toString().padStart(3, '0')
    }

    fun addNote(note: Note) {
        notes.add(note)
    }

    fun noteById(noteId: String): Note? = notes.firstOrNull { it.id == noteId }

    fun replaceNote(note: Note): Boolean {
        val index = notes.indexOfFirst { it.id == note.id }
        if (index < 0) return false
        notes[index] = note
        return true
    }

    fun removeNote(noteId: String): Boolean = notes.removeAll { it.id == noteId }

    fun toJson(): JsonValue = JsonValue.obj(
        "version" to JsonValue.num(version),
        "manuscript_path" to JsonValue.str(manuscriptPath),
        "manuscript_hash" to JsonValue.str(manuscriptHash),
        "position" to JsonValue.obj(
            "chapter_id" to JsonValue.str(position.chapterId),
            "sentence_index" to JsonValue.num(position.sentenceIndex),
        ),
        "notes" to JsonValue.arr(notes.map { it.toJson() }),
        "lexicon" to JsonValue.arr(lexicon.map { it.toJson() }),
    )

    fun toJsonString(): String = toJson().toJsonString() + "\n"

    fun copyOf(): Sidecar = Sidecar(
        version = version,
        manuscriptPath = manuscriptPath,
        manuscriptHash = manuscriptHash,
        position = position,
        notes = notes.toMutableList(),
        lexicon = lexicon.toList(),
    )

    companion object {
        fun new(manuscriptPath: String, manuscriptHash: String): Sidecar = Sidecar(
            manuscriptPath = manuscriptPath,
            manuscriptHash = manuscriptHash,
        )

        fun fromJson(value: JsonValue): Sidecar {
            val position = value.objectAt("position")
            return Sidecar(
                version = value.int("version", SIDECAR_VERSION),
                manuscriptPath = value.string("manuscript_path"),
                manuscriptHash = value.string("manuscript_hash"),
                position = Position(
                    chapterId = position.string("chapter_id"),
                    sentenceIndex = position.int("sentence_index"),
                ),
                notes = value.array("notes").map { Note.fromJson(it) }.toMutableList(),
                lexicon = value.array("lexicon").map { LexiconRule.fromJson(it) },
            )
        }

        /** Parse sidecar text, falling back to a fresh sidecar when the file is unreadable. */
        fun parse(text: String?, manuscriptPath: String, manuscriptHash: String): Sidecar {
            val value = text?.takeIf { it.isNotBlank() }?.let { JsonValue.parse(it) }
                ?: return new(manuscriptPath, manuscriptHash)
            val sidecar = fromJson(value)
            sidecar.manuscriptPath = manuscriptPath
            return sidecar
        }
    }
}

fun sidecarName(manuscriptName: String): String =
    manuscriptName.substringBeforeLast('.', manuscriptName) + ".ultidraft.json"

fun nowIsoSeconds(): String =
    OffsetDateTime.now().truncatedTo(ChronoUnit.SECONDS).toString()

/**
 * Fold phone edits into whatever the sidecar on disk says now.
 *
 * Syncthing can land a desktop version of the file between our read and our write, so
 * every save re-reads and merges instead of overwriting: notes are unioned by id, the
 * local copy wins for ids present in both, and disk order is preserved so the export
 * does not shuffle on every save.
 */
fun mergeSidecars(disk: Sidecar, local: Sidecar): Sidecar {
    val merged = disk.copyOf()
    merged.manuscriptPath = local.manuscriptPath
    merged.manuscriptHash = local.manuscriptHash
    merged.position = local.position

    val localById = local.notes.associateBy { it.id }
    for (index in merged.notes.indices) {
        val replacement = localById[merged.notes[index].id]
        if (replacement != null) merged.notes[index] = replacement
    }
    // Anything the desktop deleted while we held a copy stays deleted; only ids we
    // actually created or edited on the phone come back.
    val diskIds = disk.notes.map { it.id }.toSet()
    for (note in local.notes) {
        if (note.id !in diskIds && note.id.startsWith(MOBILE_NOTE_PREFIX)) merged.notes.add(note)
    }

    val lexicon = merged.lexicon.toMutableList()
    for (rule in local.lexicon) {
        if (lexicon.none { it.written.equals(rule.written, ignoreCase = true) }) lexicon.add(rule)
    }
    merged.lexicon = lexicon
    return merged
}
